### PR TITLE
ENH - use sparse matrices in ``process_n_gram``

### PR DESCRIPTION
**Pull request by basanthsk**
I optimized the ngram computation; now it takes 30 seconds instead of more than a minute to complete.

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
Thanks @basanthsk for this amazing and promising PR!

Can you provide an example that shows the speed gain as well as the timing comparison between both versions?

**Comment by basanthsk** [comment]
Link to the test file for comparison
https://drive.google.com/drive/folders/1cCsT9DEPyyKXN5W34ETtwCkiOx1mwobh?usp=sharing

The original 
image
Optimizer
image

**Comment by Badr-MOUFAD** [comment]
Congratulation @basanthsk for this :rocket:  PR!
I played around with the provided examples and it is super fast :zap:.

Before merging, we shall make minor changes to the code for readability and future maintenance.
I will make a code review very soon.

**Comment by basanthsk** [comment]
Thanks @Badr-MOUFAD
Should I do the changes/you will do it/we will doing together?

File: requirements.txt [hunk 1]
```diff
@@ -6,4 +6,5 @@ jellyfish
 sphinx
 furo
 sphinx-copybutton
-twine
\ No newline at end of file
+twine
+scipy
```
Comment (Badr-MOUFAD) [inline requirements.txt:10]: should also be added to dependencies in ``setup.py``

File: skillNer/utils.py [hunk 2]
```diff
@@ -43,17 +44,20 @@ def grouper(self, iterable, dist):
 
     def get_clusters(self, co_oc):
         clusters = []
-        for i, row in enumerate(co_oc):
+        i=0
+        for row in co_oc.tolil().rows:
```
Comment (Badr-MOUFAD) [inline skillNer/utils.py:48]: we better refactor this ``for`` loop:
1. handle the opposite case
```python
for ...:
    if len(row) == 0:
        continue
    # the rest of implementation
```
2. use enumerate co_oc.tolil().rows

File: skillNer/utils.py [hunk 3]
```diff
@@ -73,21 +77,25 @@ def get_corpus(self, text, matches):
...
         unique_skills = list(set([match['skill_id'] for match in matches]))
         skill_text_match_bin = [0]*len_
-        for index, skill_id in enumerate(unique_skills):
+        match_df = pd.DataFrame(matches)
+        match_df_group = match_df.groupby('skill_id')['doc_node_id']
+        corpus=[]
+        look_up = {}
+        idx=0
+        on_inds =[]
+        for skill_id,g in match_df_group:
```
Comment (Badr-MOUFAD) [inline skillNer/utils.py:90]: same here, we better use ``enumerate``

**Review by Badr-MOUFAD — COMMENTED** [review]
We shall do the changes together @basanthsk

here is 1st round review. Pin me for any remarks.

File: requirements.txt (hunk 1 above)
Comment (basanthsk) [reply requirements.txt:10]: yes

File: skillNer/utils.py (hunk 2 above)
Comment (basanthsk) [reply skillNer/utils.py:48]: done the changes

File: requirements.txt (hunk 1 above)
Comment (basanthsk) [reply requirements.txt:10]: done

**Comment by basanthsk** [comment]
@Badr-MOUFAD I have done the chages you requested also tested

**Comment by Badr-MOUFAD** [comment]
thanks @basanthsk but I can't see the changes, you may have forgotten to push

**Comment by basanthsk** [comment]
@Badr-MOUFAD  my bad I pushed it to one of the branch in my repo
Now I have merged it into main brach
Please have look 
Do I have to make one more pull request after changes for review?

**Comment by Badr-MOUFAD** [comment]
> Do I have to make one more pull request after changes for review?

No need

Thanks again @basanthsk for this work :muscle:. These improvements would take effect after a new release.
@AnasAito, pls have a look and merge if it's :+1:

**Pull request merged**